# Basic types and operations

## $ \S 1 $ Objects, classes, values, variables and assignment

An __object__ is an instance of a specific data type that resides in memory. In
Python, everything is represented as an object, and each object belongs to a
__class__, which determines how it is implemented, its properties and the
operations that can be applied to it. Each object created by the Python
interpreter is assigned a unique internal identifier (which we do not need to
handle directly or worry about). The actual data or information that an object
holds is referred to as its __value__.

When a class is already built into Python, it is usually called a __type__. For
example, there are types for the representation of integers, text strings or
boolean values. We can divide an integer by another, but not a string by
another; we can concatenate two strings, but that doesn't make sense for boolean
values.

On the other hand, in writing software for a bank, for instance, it would be
extremely cumbersome to work directly only with these built-in types. In order
to encapsulate the data and behaviors associated with the manipulation of bank
accounts, it would be better to define an appropriate class to represent them at
a higher-level.

📝 In the terminology of object-oriented programming, the technical terms for
"property" and "operation" are _attribute_ and _method_, respectively. Also,
an object belonging to a class is said to be an _instance_ of that class.

A __variable__ is a symbolic name that points to an object. Variables
allow us to interact with objects in a more intuitive and convenient manner,
without needing address them by their specific memory locations. Variables are
created through an __assignment__ statement, using the __assignment operator__
`=` in the form _\<variable name\>_ `=` _\<object>_.  This binds the object to
the right of the `=` sign to the label to its left.

In [14]:
a = 1729   # Bind the name 'a' to the integer 1729.
a          # Return the value of the corresponding object as output of the cell.

📝 Unlike some other programming languages, in Python the syntax for _creating
a new variable_ and for _assigning a new object to an existing variable name_ is
exactly the same:

In [2]:
b = 1729       # Create a new variable b and assign the object 1729 to it.
print(b)       # Print the current value of b.
b = 17.29      # Reassign a _new_ object to the name 'b'.
print(b)       # Print the current value of b.

1729
17.29


Variable names may consist only of __alphanumeric characters__ (i.e., the
characters A-Z, a-z and 0-9) and __underscores__ \_; the initial character must be
a letter. By convention, uppercase initial letters are usually reserved to
denote classes and lowercase initial letters to name other objects. 

In this tutorial we will not need to discuss objects, classes and object-oriented
programming beyond what has already been mentioned.

In [1]:
wiz30 = 'Presto'    # Assign the string 'Presto' to the name 'wiz30'.
wiz30               # Return the value of variable w as the output of the cell.

'Presto'

<div class="alert alert-info"> In Python, simply creating a variable or assigning it a new object does not prompt the interpreter to return nor print its value.
    <ul>
        <li>To get the interpreter to <i>return</i> the value of the object
        bound to a variable, say $ x $, as the output of a cell, one can use a
        statement consisting solely of its name, in this case <code>x</code>.</li>
        <li>To <i>print</i> the value of the object referred to by a variable,
        say $ x $, to the screen, use <code>print(x)</code>. Note that calling
        <code>print</code> <i>generates no output</i> (more precisely, it
        returns <code>None</code>).</li></ul></div>

More generally, the function `print` can be used to print to the screen any collection of arguments, separated by commas. These commas will be replaced by single spaces when the arguments are displayed.

In [2]:
a = 'cunning'
b = 'wizard'
print(a)
print(a, b)

cunning
cunning wizard


⚠️ If a code cell contains more than one statement which produces an output,
then only the result of the most recent one will be yielded as the output of the
cell; the other ones are discarded. 

In [10]:
# This statement causes the interpreter to evaluate the variable a:
a    # (evaluates to the string 'cunning')
# However, if we issue another such statement:
b    # (evaluates to the string 'wizard')
# then only the latter will actually yield an output.

'wizard'

The type of a variable $ x $ can be inspected through a call to the function
`type`, as in `type(x)`. 

In [1]:
number = 'two'
# The string 'two' is assigned to the new variable number.
print(number)
print(type(number))

number = 2
# The variable number now points to an object of different type.
print(number)
print(type(number))

two
<class 'str'>
2
<class 'int'>


As in the preceding example, the Python interpreter automatically infers the
type of a variable from its value; this feature is called __type inference__.
Moreover, in Python (in contrast to, say, C or Java), variables are
__typed dinamically__, meaning that not only can the _object_ bound to a
variable change during a program, but even its _class_ (or _type_) can be
modified. Whenever a new object is assigned to a previously used name, the
original association is lost. 

📝 Blank lines in a code cell are ignored by the interpreter. We can use them to
visually separate the code into several parts in order to improve legibility.

The statement `isinstance(<variable>, <class>)` returns either `True` or `False`
according to whether a given variable belongs to the indicated class or not.

__Exercise:__ If the statement `lifespan = 120` is run, determine the
values of:

(a) `isinstance(lifespan, str)`

(b) `isinstance(lifespan, int)`

(c) `isinstance(lifespan, float)`

How do the answers change if `lifespan = 120.0`?

In [2]:
lifespan = 120

⚠️ All names (not just those of variables) are __case-sensitive__. Thus `Staff` and `staff` may refer to completely different objects.

__Exercise:__ Predict what gets printed when you run the code cell below.

In [ ]:
Mammal = "whale"
print(Mammal)
mammal = "dog"
print(mammal)
print(Mammal)

__Exercise:__ Can you explain what happens if the statement `print(print(a))`is interpreted?

__Exercise:__ What happens if you issue the following statements consecutively?
What is the output (if any) in each case?

(a) `a = 2`

(b) `b = 3`

(c) `c = a * b`

(d) `print(c)`

(e) `a - b`

(f) `a**b`

🚫 The following small set of keywords *cannot* be used to name objects because they already have a special meaning to the interpreter:

| Reserved  | keywords   |            |           |
| :-------- | :--------- | :--------- | :------   |
|`False` 	| `def` 	 | `if`       | `raise`   |
|`None` 	| `del` 	 | `import`   | `return`  |
|`True` 	| `elif` 	 | `in` 	  | `try`     |
|`and` 	    | `else` 	 | `is` 	  | `while`   |
|`as` 	    | `except`   | `lambda`   | `with`    |
|`assert` 	| `finally`  | `nonlocal` | `yield`   |
|`break` 	| `for` 	 | `not`      |           |
|`class` 	| `from` 	 | `or` 	  |           |
|`continue` | `global`   | `pass`     |           |


__Exercise:__

(a) Suppose `x` and `y` have values $ 1 $ and $ 2 $, respectively. If we want to
swap these values, can we accomplish that through the instructions in the code
cell below? Explain.

In [3]:
x = 1    # Initializing x.
y = 2    # Initializing y.

# Performing the swap:
x = y
y = x
print(x, y)

(b) How could one use a third, temporary variable `temp` to solve this problem?

In Python, one can make several __simultaneous assignments__ in a single line
using commas `,` as separators. This is especially useful for permuting the
values of two or more variables without resorting to temporary variables,
as in the preceding problem.  However, as a best practice one should avoid
needlessly using this feature because it hampers the legibility of code.

__Example:__

In [45]:
x = 1
y = 2
print(x, y)

x, y = y, x
print(x, y)

1 2
2 1


In [6]:
# Multiple assignments are made _simultaneously_.
# Check this in the following example:
x = 1
y = 2
x, y = x + y, x - y

__Exercise:__ As we have said before, in Python every object must belong to
a class, i.e., be of some type. How could you find out what the type of a given
type is? Does the answer depend on the given type?

## $ \S 2 $ The boolean type `bool` 

### $ 2.1 $ Description of the boolean type

The __boolean__ type, denoted by `bool` in Python, consists of only two objects:
`True` and `False`.

Booleans support the three basic __logical operators__ `and`, `or` and `not`
(formally called _conjunction_, _disjunction_ and _negation_, respectively).
Observe that the former two are _binary_ operators (i.e., requiring two
arguments), whereas the latter is a _unary_ operator (i.e., it works on a single
boolean argument).

A __boolean expression__ is an expression that has a boolean value, that is,
which evaluates to either `True` or `False`. 

Booleans are extremely important in any programming language because they enable
the conditional execution of pieces of code. They usually appear in the
conditional tests of `if-elif-else` or `while` constructs to be considered later,
or as the result of comparisons. However, they can be useful in several other
situations as well.

__Examples:__

In [48]:
a = True
b = False
print(a, type(a))
print(b, type(b))

True <class 'bool'>
False <class 'bool'>


__Exercise:__ By considering all pairs that can be formed from the values `True`
and `False`, construct the truth tables of the `and` and `or` operators. _Hint:_
To reduce the amount of text that needs to be typed, introduce two variables
`t` and `f` that have `True` and `False` as their values.

__Exercise:__ Let $ a $ and $ b $ be variables having the values `True` and
`False`, respectively. What is the value of the following boolean expressions?

(a) `not(a)`

(b) `not a`

(c) `not(a and b)`

(d) `not a and b`

(e) `(not a) or (not b)`

(f) `not a or not b`

(g) `not not a`

(h) `a and not b`

(i) `a and b or a`


### $ 2.2 $ Comparison operators <a name="comparison"></a>

The following binary comparison operators can be applied to several types of
objects. Each of them yields either `True` or `False` as output.

| Operator   | Translation                |
| :--------  | :---------                 |
| `==`       | Equal to                   |
| `!=`       | Not equal to               |
| `<`        | Less than                  |
| `>`        | Greater than               |
| `<=`       | Less than or equal to      |
| `>=`       | Greater than or equal to   |


<div class="alert alert-warning">It is a common mistake for beginners to confuse the assignment operator <code>=</code> with the equality operator <code>==</code>; this can lead to syntactical or, worse, semantical errors.</div>

📝 If a comparison is made between two values of different types, then the interpreter will first try to convert them to a common type. For example, if one compares an integer to a float, then the integer is first converted to a float.

__Exercise:__ Predict and explain the output of the following statements:

(a) `0 == 0`

(b) `1 > 2`

(c) `1 != 0.9999999`

(d) `1 == 0.99999999999999999`

(e) `2 == 1 + 1`

(f) `2 = 1 + 1`

(g) `True != False`

(h) `3**5 >= 4**4`

(i) `"aba" < "abc"`

__Exercise:__ Let $ a = 8 $, $ b = 7.99 $ and $ c = -23 $. What do the following boolean expressions evaluate to?

(a) `b > a`

(b) `a != b`

(c) `a == a`

(d) `a >= a`

(e) `c <= a and a >= b`

(f) `c < b < a `

(g) `c != b > a`

(h) `a <= a >= a`

In [16]:
a = 8
b = 7.99
c = -23

📝 In Python, any value of a built-in type can be interpreted as a Boolean. 
As a rule of thumb, for numerical types (`int`, `float`, `complex`), $ 0 $
is the only number that evaluates to `False`. For sequential types
(such as `str`, `list` or `tuple`), only empty objects are considered `False`.
In other words, `True` and `False` are not the only objects considered
true and false!  We can inspect how a value will be interpreted as a Boolean by
explicitly converting it to the type `bool`:

In [1]:
print(bool(0))
print(bool(1 + 3j))
print(bool(""))
print(bool("Das sind Geschichten"))

False
True
False
True


## $ \S 3 $ `None`

The type `Nonetype` consists of the single object `None`, which is used to
indicate nothing, i.e., a null value. It is frequently used as a placeholder in
the case where a function has no values to return or to indicate that some
object is empty or missing. For example, a call to the `print` function always
returns `None` as its output, regardless of its argument.

* `None` is not the same as $ 0 $; it is not possible to use it
  within arithmetic expressions.
* `None` is not a string, empty or otherwise. However, if it is printed, 'None'
  is displayed.
* `None` is not the same as `False`, but it is evaluated to `False` when appearing
  in a conditional test.

__Exercise:__ Let `x = None`. Check the output of the following statements
(trying to explain some of them would require more knowledge of Python than we
have at this point):

(a) `x`

(b) `print(x)`

(c) `print(print(x))`

(d) `x == True`

(e) `x == False`

(f) `x != 0`

(g) `not None`

In [2]:
x = None

## $ \S 4 $ Numerical types and operators

Python supports three built-in data types for representing numbers:
* `int`, or __integer__ type, for integers such as $ -1 $, $ 2 $, $ 0 $ or $ 53 $;
* `float`, or __floating-point__ type, for floating-point numbers (intuitively,
  numbers with a finite decimal expansion) such as $ 3.1415 $, $ 2.0 $ or
  $ -.450 $;
* `complex`, or __complex__ type, for complex numbers such as $ 2 + 3i $ or $
  3.14 - 43.5 i$.

Each of these types also supports the equality `==` and inequality `!=`
operators, the basic arithmetic operators `+`, `-`, `*` (multiplication), `/`
(division) and a few others to be discussed below. Integers and floats (but not
complex numbers!) can also be compared in size through `<`, `>`, `<=` and `>=`.

## $ \S 5 $ The type `int` of integers

Unlike in many other languages, there is only one type for representing
integers: `int`. Moreover, these integers can in principle be of any size (a
limit is imposed only by the memory capacity of the computer).

In [3]:
a = 4
b = -3
print(a, type(a))    # Checking if a is of type int.
print(b, type(b))

4 <class 'int'>
-3 <class 'int'>


In [1]:
# Verifying if the author lied about there being no bound to
# the possible values of integers:
print("2 raised to the 64:", 2**64)
print("2 raised to the 256:", 2**256)
print("2 raised to the 1024:", 2**1024)

2 raised to the 64: 18446744073709551616
2 raised to the 256: 115792089237316195423570985008687907853269984665640564039457584007913129639936
2 raised to the 1024: 179769313486231590772930519078902473361797697894230657273430081157732675805500963132708477322407536021120113879871393357658789768814416622492847430639474124377767893424865485276302219601246094119453082952085005768838150682342462881473913110540827237163350510684586298239947245938479716304835356329624224137216


__Exercise:__ Can you predict the value of the expressions below when $ x = 3 $ and $ y = 2 $? What if $ y = - 2 $?

In [ ]:
# x = 3
# y = 2
print(x + y)      # Adding x and y.
print(x - y)      # Subtracting y from x.
print(x * y)      # Multiplying x and y.
print(x // y)     # Quotient of x by y (integer division).
print(x % y)      # Remainder of the division of x by y.
print(y**x)       # Raising y to x; may return a float.
print(x**y)       # Raising x to y; may return a float.
print(x / y)      # Division of x by y; always returns a float.

__Exercise:__ Let $ x = 64 $. What do the statements `x**0.5` and `x**(1/3)` yield? Explain.

__Exercise:__ What is $ 0^0 $ according to Python? 

## $ \S 6 $ The type `float` of floating-point numbers

Irrational numbers such as $ \pi $ or $ e $ can never be represented exactly on
a machine, no matter whether the decimal or binary system is used,
because this would require an infinite amount of digits to be stored.
Floating-point numbers provide approximate representations of real numbers;
their type is called `float`. Using the conventional notation, a floating-point
number is characterized by the use of a decimal point `.` to separate its
integer and fractional part.

__Example:__

In [45]:
x = 3.14         # We can recognize that x and y are of type
y = -2.71        # float because of the decimal points.
z = 19.
w = .456
print(type(z))   # Verifying that z is of type float.
print(x**y)      # Raising x to y.
print(x / x)     # Dividing x by x.

<class 'float'>
0.04501120645942232
1.0


📝 Following the IEEE 754 standard, in Python floating-point numbers are
represented as $ 64 $-bit, double-precision values. The values that any
floating-point number can take on are restricted to the approximate range
between $ \pm 1.8 \cdot 10^{308} $.  If some expression yields a value that
exceeds these bounds, then an _overflow error_ or some other unexpected quantity
may result. Similarly, a nonzero number whose absolute value is smaller than $
2.3 \cdot 10^{-308} $ cannot be represented exactly.


__Exercise:__ Let $ r = 3.0 $ and $ s = 1.2 $. Guess and then explain the values
that are printed below.

In [ ]:
# r = 3.0
# s = 1.2
print(r + s)      # Adding r and s.
print(r - s)      # Subtracting s from r.
print(r * s)      # Multiplying r and s.
print(r / s)      # Division of r by s.
print(r**s)       # Raising r to s.
print(r // s)     # How many times does s fit into r? Returns a float.
print(r % s)      # And what is the remainder? Returns a float.

__Exercise:__ What happens if you try to divide by $ 0 $?

A floating-point number can also be written in the __scientific__ or __exponential notation__ using `e`. In this notation, the number to the left of `e` is multiplied with $ 10 $ raised to the power to the right of `e`.

__Exercise:__ Which real numbers are represented by the following floats?

(a) `3.14159e0`

(b) `1.23e2`

(c) `2e-1`

(d) `1.e-2`

(e) `10e-1`

(f) `4.56e-2`

(g) `.789e5`

⚠️ Note that `10e1` does not represent $ 10^1 = 10 $, but rather
$$
10 \times 10^1 = 10^2 = 100\,.
$$
In other words, the `e` in the scientific notation does not stand
for the exponentiation operation, which is instead denoted by `**`.

⚠️ Scientific notation can only be used with constant values, not variables. 

__Exercise:__ Let `x = 3.14` and `y = 2`. Consider the following three ways to
get the interpreter to compute the value of $ 3.14 \times 10^2 $ and explain
their results:

(a) `print(3.14e2)`

(b) `print(xe2)`

(c) `print(3.14ey)`

📝 The expression "floating-point" refers to the fact that, when expressed
in scientific notation, the position of the decimal point of a number of this
type can "float" anywhere between its digits, i.e., it is not bound to the
position to the right of the digit representing units. In particular, the
same number can be represented in several ways, e.g.:
$$
3.1415 \times 10^0 = 3141.5 \times 10^{-3} = 31415 \times 10^{-4}
$$

__Exercise:__ Explain why `2023**100` evaluates to an integer while `2020.1**100` yields an overflow error.

__Exercise:__ Evaluate `2.0 + 2.0 - 2.0` and `2.0 + 2.0e16 - 2.0e16` and try to explain why the results are different. (We will return to this problem when we study floating point arithmetic later.)

## $ \S 7 $ The type `complex` of complex numbers

We will rarely need to use complex numbers in this course. A number of type
`complex` can be represented in Python as a pair of floats (its real and
imaginary parts) using the special notation indicated in the following examples.

⚠️ In Python the imaginary unit is denoted by $ j $ instead of the more usual
$ i $. 

In [77]:
w = 1.23 + 4.56j  # Note the syntax: we write '4.56j', not '4.56 * j'.
print(w, type(w))    # Checking the type of w.

z = 4 + 1j       
print(z, type(z))

print(z + w)

(1.23+4.56j) <class 'complex'>
(4+1j) <class 'complex'>
(5.23+5.56j)


⚠️ By itself, `j` gets interpreted as a variable whose name is $ j $.  In order
to indicate that we want a complex number, `j` must be immediately preceded by a
number.

In [118]:
print(j)    # Will result in an error if there is no variable named 'j'.

NameError: name 'j' is not defined

In [ ]:
# Even if the imaginary part is zero, it must still be included explicitly to
# indicate that we are dealing with a complex number instead of a float
# or an int:
a = 1 + 0j
print(a, type(a))
# However, in case the real part is 0, we may omit it:
b = 1j
print(b, type(b))

(1+0j) <class 'complex'>
1j <class 'complex'>


📝 Each complex number is provided with special attributes `real` and `imag` to
representing its real and imaginary parts respectively, and a special method
`conjugate` to compute its conjugate. 

In [79]:
# Complex number with real part 0 and imaginary part 1.0:
w = 1.0j
print(w)    
print(w.real, type(w.real))
print(w.imag, type(w.imag))

1j
0.0 <class 'float'>
1.0 <class 'float'>


In [80]:
# Complex number with real part 0 and imaginary part 1:
z = 0 + 1j
# Printing z and the type of its real and imaginary parts:
print(z)    
print(z.real, type(z.real))
print(z.imag, type(z.imag))
# Although we used the integers 0 and 1 as the real and imaginary parts, the
# resulting complex number still has real and imaginary parts of type float!

1j
0.0 <class 'float'>
1.0 <class 'float'>


In [81]:
print(z.conjugate(), type(z.conjugate()))

-1j <class 'complex'>


__Exercise:__ Suppose that the statements `z = 1 + 1j` and `w = 1 -1j` have just been interpreted.
What is the value and type of the following expressions?

(a) `z + w`

(b) `z - w`

(c) `z**2`

(d) `z**4`

(e) `z * w`

(f) `z / w`

(g) `w != z`

(h) `z + w <= 2 * (z + w)`

__Exercise:__ Use Python to guess the value of $ e^{\pi i} $.

## $ \S 8 $ Type conversion

We can use `int`, `float` and `complex` as _functions_ to convert a value to
the corresponding type in certain cases. Similarly, any value of a numerical
type can be converted to a string by using `str` as a function (strings will be
discussed in depth in the next notebook). More precisely:

* Any integer may be converted into a floating-point or complex number;
* Any floating-point number can be converted into a complex number;
* Any numerical value can be converted into a string;
* Conversely, a string can also be converted into a numerical type, provided
  that its literal represents a valid number of the intended type.

__Exercise:__ Suppose that $ a = 2 $. If we run the following statements in
sequence, what are the values and types of the corresponding outputs?

(a) `float(a)`

(b) `complex(a)`

(c) `str(a)`

(d) `a`

📝 Note that in any of these cases, the original object remains the same and a
_new_ object having the specified type is created. For instance, in the
preceding exercise, after all the statements have been run, $ a $ is still of
type `int`.

Applying `int` to a positive floating-point number truncates its decimal part,
that is, returns the largest integer $ n \le x $, or rounds $ x $ _down_. If
$ x $ is negative, however, then `int(x)` returns the smallest integer $ n \ge x
$, i.e., rounds $ x $ _up_.

__Exercise:__ Compute the outputs of:

(a) `int(2)`

(b) `int(2.3)`

(c) `int(2.8)`

(d) `int(-2.3)`

(e) `int(-2.8)`

To round $ x $ to the nearest integer, we use `round(x)`. More generally,
`round(x, <number of digits>)` rounds a floating-point number $ x $ to a precision
within the specified number of digits after the decimal point.

__Exercise:__ Let the variable $ b $ have the value $ 2.654321 $. What are the values of:

(a) `int(b)`

(b) `round(b)`

(c) `round(b, 3)`

(d) `round(b, 0)`

(e) `round(b, 9)`

(f) `round(-b)`

(g) `int(-b)`

In [103]:
b = 2.654321

🚫 Trying to convert a complex number into an integer or floating-point number
will result in a `TypeError`.

__Exercise:__ Suppose that the statement `z = 2.0 + 0j` has just been run through the interpreter. Describe the outputs of:

(a) `int(z)`

(b) `float(z)`

(c) `str(z)`

(d) `complex('2.0 + 0j')`

(e) `complex('2.0')`

## $ \S 9 $ Arithmetic operations
In summary, Python supports the following binary arithmetic operators:

| Operator  | Operation          | Types allowed              |
| :-------- | :---------         | :------------------------  |
| `+`       | Addition           |  `int`, `float`, `complex` |
| `-`       | Subtraction        |  `int`, `float`, `complex` |
| `*`       | Multiplication     |  `int`, `float`, `complex` |
| `/`       | Division           |  `int`, `float`, `complex` |
| `**`      | Exponentiation     |  `int`, `float`, `complex` |
| `//`      | Integer division   |  `int`, `float`            |
| `%`       | Modulo (remainder) |  `int`, `float`            |

<a name="table 1"></a>

The latter two operators are defined as follows. If $ x $ and $ y $ are
_positive_ integers or floating-point numbers, then:
* ` x // y` is the largest integer $ n $ such that $ n y \le x $.
* `x % y` is the remainder of `x // y`, that is, $ x - n y $ for $ n $ as above.

These operators are usually applied only when $ x $ and $ y $ are both positive. If one or both of $ x $, $ y $ is negative, the definition is similar.

📝 Note the similarity of the symbol `%` to the division operator $ \div \,$.

__Exercise:__ Let $ x = 11 $, $ y = 3 $, $ s = 2.51 $ and $ t = 8.53 $. What is the output of the statements below?

(a) `x // t`

(b) `x % t`

(c) `t // y`

(d) `t % y`

(e) `t // s`

(f) `t % s`

__Exercise:__ By considering some examples, deduce the meaning of `x // y` and `x % y ` when one (or both) of $ x $ and $ y $ is negative.

📝 Applying `%` with second argument $ 1 $ results in the fractional part of the
(positive) first argument, as in the following example.

In [108]:
3.14159 % 1

0.14158999999999988

__Exercise:__ Predict the values of the following expressions and then check your answers:

(a) `7 * 2`

(b) `7 ** 2`

(c) `7 // 2`

(d) `7 / 2`

(e) `7 % 2`

Besides these binary arithmetic operators, we can also make use of the following functions when working with numerical values:
* `abs`, the **absolute value** function, which takes a single numerical argument of any numerical type (`int`, `float` or `complex`) and returns its absolute value (or modulus, in the case of complex numbers);
* `max` and `min`, which can take any number of arguments of type
  `float` or `int` separated by commas and return their __maximum__ and
  __minimum__, respectively.

__Exercise:__ Determine the output of the following statements:

(a) `abs(-2.71828)`

(b) `abs(2)`

(c) `abs(-3 + 4j)`

(d) `max(-5.3, 2, 10.45, -23, 0)`

(e) `min(-5.3, 2, 10.45, -23, 0)`

(f) `max(1 + 1j, 2 + 1j)`


Other common mathematical functions such as the exponential, the logarithm and the trigonometric functions can be used after importing the `math` or `numpy` modules (more about this later).

## $ \S 10 $ Compound assignment operators

Corresponding to each of the arithmetic operators considered above, Python
provides the following __compound assignment operators__ which can be used to
perform an operation in-place on a variable provided as its first operand:

    +=, -=, *=, /=, **=, //=, %=

Their meaning can be gleaned from the following examples.

| Example             | Equivalent statement |
| :--------           | :---------           |
| `x += 3.14`         | ` x = x + 3.14`      |
| `x -= (2  + 3j)`    | `x = x - (2 + 3j)`   |
| `x *= 2`            | `x = x * 2`          |
| `x /= 3`            | `x = x / 3`          |
| `x **= 0.5`         | `x = x**0.5`         |
| `x //= z`           | `x = x // z`         |
| `x %= y**2`         | `x = x % (y**2)`     |

In each case, the _current_ value of the variable $ x $ is used to compute
the expression on the right, and the result is then assigned as the new 
value of $ x $.

__Exercise:__ Let $ x = 7 $ and $ y = 5 $. Determine the
values of $ x $ and $ y $ after each of the following statements is run through
the interpreter in sequence:

(a) `x -= y`

(b) `y += x`

(c) `y **= x`

(d) `y //= x`

(e) `y %= x`

(f) `x **= 1/2`

In [ ]:
x = 7
y = 5